tm2/pkg/bft/rpc/lib/client/http/client.go

```diff
@@ -135,17 +135,20 @@ func sendRequestCommon[T requestType, R responseType](
 	}
 	defer httpResponse.Body.Close() //nolint: errcheck
 
-	// Parse the response code
-	if !isOKStatus(httpResponse.StatusCode) {
-		return nil, fmt.Errorf("invalid status code received, %d", httpResponse.StatusCode)
-	}
-
-	// Parse the response body
+	// Read the response body
 	responseBytes, err := io.ReadAll(httpResponse.Body)
 	if err != nil {
 		return nil, fmt.Errorf("unable to read response body, %w", err)
 	}
 
+	// Parse the response code
+	if !isOKStatus(httpResponse.StatusCode) {
+		return nil, fmt.Errorf(
+			"invalid status code received, %d. Response body: %q",
```

Comment:
```suggestion
			"invalid status code received, %d. Response body:\n%s",
```

what do you think?

maybe also `strings.ReplaceAll(string(responseBody), "\n", "\n\t")`.

Reply:
We actually lost the display of server errors for non 2xx codes in #350. 
Had another branch where tried to decode the jsonrpc ABCIError if possible in #2631. It was embarrasingly long  so prepared this short PR instead, hence `%s` or `%q`. 

`%s` kind of tries to format things, with `\n` being shown on many lines. But if it's about showing nice things, #2631 was nicer as it decoded the ABCIError. If it's about showing raw stuffs from the server in case of a 500 (a rare occurence), I think `%q` does the job. In the end it's mostly the same, there is perhaps just less ambiguity with `%q`, everything being packed together in an unmistakably compact way (with ugly \" quotes).

What is best for users? I think seeing the ABCIError from the server is the most important thing (it's valuable as it saves time). 5xx is supposed to be pretty rare.

If you're comfortable with `%s` we can use `%s`. I perhaps lack the notion of how frequently those messages will be shown.

Reply:
Hey. Sorry for the slow turnaround here. I needed to double-check some information, and I lost track of this discussion.

The thing is, most often the users of the rpcclient understand how to handle errors and print useful information. An example is the keys client itself, which will print the logs I'm familiar with and I'm pretty sure you are too:

https://github.com/gnolang/gno/blob/2e56ecf818cb8bf93d88f637757c63eeddfd8f91/tm2/pkg/crypto/keys/client/maketx.go#L206-L216

So, it seems that for the RPC server to return a non-200, and thus have an error here, there needs to be a specific condition. My research has turned up this specific code, which seems to line up with the error you're seeing:

https://github.com/gnolang/gno/blob/2e56ecf818cb8bf93d88f637757c63eeddfd8f91/tm2/pkg/bft/rpc/lib/server/http_server.go#L142-L175

Now, this kind of error will happen quite exceptionally, as in general the panic must escape the handlers and only be caught into this point.

So, I think we can have a simple handling for any 500 return here, where we simply try to parse the value of `.error.data` (can be in a simple `struct{ Error struct { Data string `json:"data"` } `json:"error"` }`, and if that unmarshaling succeeds add it to the error message.

Other cases where there are errors in the RPC handler generate an RPC response with 200 status code, and so the client won't return an error (although it will be contained within the resp).

Reply:
> So, I think we can have a simple handling for any 500 return here, where we simply try to parse the value of .error.data (can be in a simple struct{ Error struct { Data string json:"data"}json:"error" }, and if that unmarshaling succeeds add it to the error message.

Is this understanding correct: 
- Remove the possibility of the rpc server to send any 500 code altogether (following the idea you provided),
- discard current fix.

I can handle it, but I'll need you to suggest a concise Git commit title for this as a confirmation ;)

Reply:
>Remove the possibility of the rpc server to send any 500 code altogether (following the idea you provided),

Mhh. Could be an idea. Don't know.

I guess one counter-argument is that the server sends this response in the recover, so it may send it even in the case of a batch JSON-RPC request.

So I'm more inclined towards updating this existing PR and handling the error, extracting `.error.data` and putting it in the error message.